src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy

```diff
@@ -0,0 +1,251 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle
+
+import com.google.common.collect.Maps
+import org.gradle.api.*
+import org.gradle.api.file.FileCollection
+import org.gradle.api.file.SourceDirectorySet
+import org.gradle.api.internal.file.DefaultSourceDirectorySet
+import org.gradle.api.internal.file.FileResolver
+import org.gradle.api.plugins.ExtensionContainer
+import org.gradle.api.tasks.TaskContainer
+import org.gradle.api.tasks.util.PatternSet
+import org.gradle.internal.reflect.Instantiator
+import org.gradle.internal.service.ServiceRegistry
+import org.gradle.language.base.LanguageSourceSet
+import org.gradle.language.base.internal.SourceTransformTaskConfig
+import org.gradle.language.base.internal.registry.LanguageTransform
+import org.gradle.language.base.internal.registry.LanguageTransformContainer
+import org.gradle.language.base.plugins.ComponentModelBasePlugin
+import org.gradle.language.base.sources.BaseLanguageSourceSet
+import org.gradle.language.cpp.CppSourceSet
+import org.gradle.language.cpp.internal.DefaultCppSourceSet
+import org.gradle.language.cpp.tasks.CppCompile
+import org.gradle.model.*
+import org.gradle.nativeplatform.NativeBinarySpec
+import org.gradle.nativeplatform.NativeComponentSpec
+import org.gradle.nativeplatform.NativeDependencySet
+import org.gradle.platform.base.BinarySpec
+import org.gradle.platform.base.LanguageType
+import org.gradle.platform.base.LanguageTypeBuilder
+import org.gradle.platform.base.TransformationFileType
+import org.gradle.util.CollectionUtils
+
+import javax.inject.Inject
+import java.util.concurrent.Callable
+
+class ProtobufModelPlugin implements Plugin<Project> {
+  private final FileResolver fileResolver
+
+  @Inject
+  public ProtobufModelPlugin(FileResolver fileResolver) {
+    this.fileResolver = fileResolver;
+  }
+
+  @Override
+  void apply(Project project) {
+    def gv = project.gradle.gradleVersion =~ "(\\d*)\\.(\\d*).*"
+    if (!gv || !gv.matches() || gv.group(1).toInteger() < 2 || gv.group(2).toInteger() < 7) {
+      println("You are using Gradle ${project.gradle.gradleVersion}: This version of the protobuf" +
+          " plugin for c++ requires minimum Gradle version 2.7")
+    }
+
+    // Provides the osdetector extension
+    project.pluginManager.apply('osdetector')
+
+    project.extensions.create("protobuf", ProtobufConfigurator, project, fileResolver);
+
+    project.pluginManager.apply(ComponentModelBasePlugin)
+  }
+
+  static class Rules extends RuleSource {
```

Comment:
Thanks a lot. These videos are very helpful. Let me verify my understanding:

`TaskContainer` is always needed, which will trigger `configureGeneratedSourceSets()` and `finishConfiguringTheGenerateProtoTask()` since they have `TaskContainer` as their subjects. `configureGeneratedSourceSets()`, because of `ProtobufConfigurator` as its input, will trigger `protobuf()` that creates it.

`ModelMap<NativeComponentSpec>` is needed for compiling native project, which will trigger `createDefaultProtobufSourceSets()` then `bindDefaultProtoSourceSetToDefaultCppSourceSet()`.